spec/android_unit_test_checker_spec.rb

```diff
@@ -0,0 +1,286 @@
+        }
+
+        diff = generate_add_diff(changes_dict)
+        allow(@dangerfile.git).to receive(:diff).and_return(diff)
+
+        @plugin.check_missing_tests(classes_exceptions: classes_to_ignore, subclasses_exceptions: subclasses_to_ignore)
+
+        expect(@dangerfile.status_report[:errors]).to be_empty
+      end
+
+      it 'does nothing when a PR moves code around and' do
+        shape_dict = {
+          'Shape.kt' => '"open class Polygon(sides: Int): Shape {\n  override fun draw() {\n    for (i in 1..sides) draw()\n  }\n}\n\nabstract class Shape {\n  abstract fun draw()\n}"'
+        }
+
+        polygon_test_diff_str = <<~PATCH
+          diff --git a/PolygonTest.kt b/PolygonTest.kt
+          index 8c2bed6..ae81c49 100644
+          --- a/PolygonTest.kt
+          +++ b/PolygonTest.kt
+          @@ -1,8 +1,9 @@
+          -class PolygonTest {
+          -  val sut10: Shape = Polygon(sides = 10)
+          -  val sut5: Shape = Polygon(sides = 5)
+          -
+          -  fun testDraw() {
+          -    Polygon(sides = 5).draw()
+          +class ShapesTest {
+          +  fun testPentagon() {
+          +    draw(Polygon(sides = 5))
+          +  }
+          +
+          +  private fun draw(shape: Shape) {
+          +    shape.draw()
+            }
+          }
+        PATCH
+
+        shape_diff = generate_add_diff(shape_dict)
+        polygon_diff = GitDiffStruct.new('modified', 'project/src/androidTest/java/shapes/PolygonTest.kt', polygon_test_diff_str)
+
+        allow(@dangerfile.git).to receive(:diff).and_return(shape_diff + [polygon_diff])
+
+        @plugin.check_missing_tests
+
+        expect(@dangerfile.status_report[:errors]).to be_empty
+      end
+    end
+
+    def generate_add_diff(changes_dict)
+      changes_dict.map do |file_path, content|
+        diff_str = <<~PATCH
+          diff --git a/#{file_path} b/#{file_path}
+          new file mode 100644
+          index 0000000..fd48a22
+          --- /dev/null
+          +++ b/#{file_path}
+          @@ -0,0 +1 @@
+          +#{content}
+          \\ No newline at end of file
+        PATCH
+
+        GitDiffStruct.new('new', file_path, diff_str)
+      end
+    end
+
+    def generate_deleted_diff(changes_dict)
+      changes_dict.map do |file_path, content|
+        diff_str = <<~PATCH
+          diff --git a/#{file_path} b/#{file_path}
+          deleted file mode 100644
+          index fd48a22..0000000
+          --- a/#{file_path}
+          +++ /dev/null
+          @@ -1 +0,0 @@
+          -#{content}
+          \\ No newline at end of file
+        PATCH
+
+        GitDiffStruct.new('deleted', file_path, diff_str)
+      end
+    end
+
+    def class_names_match_report?(class_names:, error_report:)
+      error_report.all? do |str|
+        class_names.any? do |class_name|
+          str.include?("Please add tests for class `#{class_name}`")
+        end
+      end
+    end
```

Comment:
I might be wrong, but I think this only checks if all errors that are reported contain the name of one of the `class_names`. But that means that the `class_names` array can have a lot more items in it and this will still be true.

```ruby
[1] pry(main)>     def class_names_match_report?(class_names:, error_report:)
      error_report.all? do |str|
        class_names.any? do |class_name|
          str.include?("Please add tests for class `#{class_name}`")
        end
      end
[1] pry(main)> def class_names_match_report?(class_names:, error_report:)
  error_report.all? do |str|
    class_names.any? do |class_name|
      str.include?("Please add tests for class `#{class_name}`")
    end
  end
=> :class_names_match_report?  
[2] pry(main)> error_report = %w[A B C].map { |c| "Please add tests for class `#{c}`" }
=> ["Please add tests for class `A`", "Please add tests for class `B`", "Please add tests for class `C`"]
[3] pry(main)> class_names = %w[A B C D E F]
=> ["A", "B", "C", "D", "E", "F"]
[4] pry(main)> class_names_match_report?(class_names: class_names, error_report: error_report)
=> true
```

Instead we need to test both directions, i.e. that not only all errors reported in `error_report` mention one of the expected `class_names`, but also that all the `class_names` are reported in `error_report`. And even further, ideally we want the list to match the counts exactly — i.e. that if `error_report` contains 4 errors but 2 of those errors are both for class `B`, then maybe we should pass `class_names: %w[A B B C]` to explicitly indicate we expect B twice.

---

Finally, personal preference but I think I'd find it easier to read at call site if that method included the `expect` call directly — as opposed to the method returning a boolean and then the call site having to do `expect(result).to be(true)` 🤷 

Maybe something like this would work then:
```ruby
expect(class_names.length).to eq(error_report.length)
class_names.zip(error_report).each do |cls, error|
  expect(error).to include "Please add tests for class `#{cls}`"
end
```

Reply:
> On this case, as we have full control over the mocked data and the test itself must be correct, I don't think this would be an issue (e.g. the person writing the test knows that the classes are the classes that should be in the diff).

The main problem is if we update the implementation in the future and accidentally break things, leading to `error_report` only reporting errors for `A` and `B` but stopping to report errors for class `C`. The unit tests that was passing `class_names: %w[A B C]` would then still pass (because `error_reports.all? { … class_names.any? … {}}` would still pass with all of `A` and `B` passing the `any?` test) and thus your test won't detect the regression.

Reply:
Ah that's true, indeed it wouldn't catch a regression like that 👍  I've already updated the code.